Full frame correlation - WIP

EirikOpheim opened the pull request:
---
name: Full frame correaltion - WIP
about: A new correlation metric that performs the comparison in Fourier space. Algorithm is found in A. Foden, D. M. Collins, A. J. Wilkinson and T. B. Britton "Indexing electron backscatter diffraction patterns with a refined template matching approach"
doi: https://doi.org/10.1016/j.ultramic.2019.112845 (Eq 3). 

What I have tried to do is set up the structure for the algorithm, to fit in with other correlation methods in pyxem. This method is way slower than the other correlation metrics, but according to the referenced article it also requires a smaller template library, at least in the case of Kikuchi patterns. Furthermore, this method depends less on centering the direct beam than "fast_correlation" and "zero_normalized_cross_correlation". 

The article suggests that the use of filters can improve results. I have not implemented any filters yet, but I have commented the locations where I believe filters can be applied. I am not quite sure how this part should look, and I hope we can discuss this below. The article suggests the use of a high-pass and a low-pass filter. Would we like smooth filtering, or a hard cut-off? I suppose user parameters should be: f_max, f_min for max and min frequency and a filter = True parameter, in case no filter should be applied.

Furthermore, a method for sub-pixel refinement of the beam center is suggested in the article. I have not implemented this method, but I believe it could be an interesting feature. 

I make use of the numpy.fft functions fft2 and iff2 functions to perform Fourier transformations. I understand that there is a discussion going on regarding best practice for Fourier transforms in Pyxem #585. I believe the code I have written here is working as intended, but there should still be room for speed up, for instance using  scipy.fft.next_fast_len as suggested by @tjof2. I am not quite sure where this should be used, though. I suppose calling this function just once per function call is optimal for speed, and that it should therefore  be called within the correlate_library function. 

---

**Release Notes**
> new feature / Work in Progress
> Summary: New correlation metric, in Fourier space. Ideal for filtering background and high frequency noise. Not so dependent on center-direct beam

**What does this PR do? Please describe and/or link to an open issue.**
A new correlation metric, taking the Fourier transform of both the template and the image, and comparing in Fourier space, as described in the referenced article. This PR is just the initial version, and I am planning to add filters and possibly also the sub-pixel refinement. In light of recent discussions on efficiency I make this PR quite early in the process of developing this feature, to make room for inputs with regard to best practice solutions. I have tested this correlation metric as is, and it has provided results of a similar quality as "fast_correlation" and "zero_normalized_cross_correlation". 

**Describe alternatives you've considered**
I have considered adding doing the scipy.fft.next_fast_len step, but I am not quite sure where this should be done. Where should the zeros be added? In the back and below the matrix?

**Are there any known issues? Do you need help?**
The code is running. I consider this a minimum working example, and plan to add filters. I am planning to add a high-pass and low-pass filter initially. If other filters should be added, please let me know. Also, I would be very grateful for some advice or an example of use for the next_fast_len function. 

**Work in progress?**
If you know there is more to do, make a checklist here:
- Add filters, find a good way to take user inputs that are understandable and easy to implement.
- Speed up code by using best practice FFT solutions.
- Write tests.
- Possibly also sub-pixel refinement if this seems to be a good idea.

tjof2 commented:
@EirikOpheim  you don't need to do the padding yourself to use `next_fast_len()`

For example, here's how you would convolve two images:

```python
import numpy as np
from scipy.fft import next_fast_len

def fft_correlation(in1, in2, normalize=False):
    """Correlation of two N-dimensional arrays using FFT.

    Parameters
    ----------
    in1, in2 : array
        Input arrays to convolve.
    normalize: bool, default False
        If True performs phase correlation.
    """
    s1 = np.array(in1.shape)
    s2 = np.array(in2.shape)
    size = s1 + s2 - 1

    # Calculate optimal FFT size
    complex_result = (in1.dtype.kind == 'c' or in2.dtype.kind == 'c')
    fsize = [next_fast_len(a, not complex_result) for a in size]

    # For real-valued inputs, rfftn is ~2x faster than fftn
    if not complex_result:
        fft_f, ifft_f = np.fft.rfftn, np.fft.irfftn
    else:
        fft_f, ifft_f = np.fft.fftn, np.fft.ifftn

    fprod = fft_f(in1, fsize)
    fprod *= fft_f(in2, fsize).conjugate()

    if normalize is True:
        fprod = np.nan_to_num(fprod / np.absolute(fprod))

    ret = ifft_f(fprod).real.copy()

    return ret, fprod
```
See how we pass the result from `next_fast_len()` straight to the `shape` argument of `fftn()`?

Also note that if your data is real-valued, you can speed things up by using `rfft` instead of `fft`.

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -132,6 +132,74 @@ def zero_mean_normalized_correlation(
+
+    #Filters can be applied here. Take user input, or standard LP/HP?
+
+    template_FT_normalization = np.sqrt(np.real(np.max(np.fft.ifft2(np.multiply(template_FT, template_FT)))))
```
You can use `template_FT * template_FT` rather than `np.multiply()` for brevity. Or even `template_FT **2`

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -132,6 +132,74 @@ def zero_mean_normalized_correlation(
+
+    template_FT_normalization = np.sqrt(np.real(np.max(np.fft.ifft2(np.multiply(template_FT, template_FT)))))
+
+    res_matrix = np.fft.ifft2(np.multiply(template_FT, image_FT)/ (template_FT_normalization * image_FT_normalization))
```
Same point about `np.multiply`

EirikOpheim replied:
Thanks for the review and the example on how to use next_fast_len! I will make the suggested changes :)

dnjohnstone commented:
Thanks for this @EirikOpheim - I think the first thing we need to do is to get clear on the various FFT correlations that we intend to have in pyxem in the short/medium term and implement an architecture that serves them all with minimal code reproduction. In this first implementation I think we should stick with straight numpy/scipty ffts but we have in our minds that in the future we intend to make the fft calculation as fast as possible - that's why we first want to maximise code re-use so that those improvements are felt everywhere in the code.

So the places we're going to need fft correlations are, the following cases:

1. Calculating the FFT correlation between a simulated diffraction pattern and a measured diffraction pattern in Cartesian coordinates (i.e. applied to a Diffraction2D object) - this should be possible at both zero image displacement and translations in x/y over a specified range - the latter being for when we need to refine the pattern center during matching. (This is the case I think you're imagining here)

2. Calculating the FFT correlation between a simulated diffraction pattern and a measured diffraction pattern in polar coordinates (i.e. applied to a PolarDiffraction2D object) - this should allow translation along the azimuthal coordinate. This will be important for comparing a library of many simulated patterns against a measured pattern because you'll get the rotation about the incident beam direction for free by doing it this way - which will mean a much smaller simulated library than if you only used (1).

3. Calculating the FFT autocorrelation between a measured diffraction pattern and itself in polar coordinates (i.e. what was introduced in #572) to assess angular symmetries in the diffraction pattern. This is effectively a special case of (2) but it might be more efficient to calculate it differently.

So what should probably happen (I think?!)  is that the _correlation() function here: https://github.com/pyxem/pyxem/blob/master/pyxem/utils/correlation_utils.py 

Is modified to support these 3 cases and called in the appropriate parts of the code to support the 3 cases mentioned.

Then after we've got that general framework in place we can worry about speed.

Does that sound right / make sense to everyone? @pc494 @CSSFrancis

CSSFrancis commented:
@EirikOpheim  This looks cool! 

**A couple of thoughts on the full frame correlation**
- One thing that could be speed up is taking the fft of the image we are trying to index only once.  This means that we don't have to calculate it over and over for each pattern.
- For the sub pixel refinement, I think this basically the same as taking the image, subdividing the pixels and then running the correlation.  If we subdivide correctly this might not be much slower if we are smart and get to a power of 2 so the FFT is fast

**For the polar coordinate matching**
@dnjohnstone Just a couple of thoughts about the pattern matching from polar coordinates. 

**Regarding what I have already implemented**

- What I have implemented only works in 1-D (the Azimuthal direction) but for the most case this is probably what you want for the polar pattern matching. The main difference is adding a reference to perform the correlation against. 
- Using the self-correlation might be an interesting step towards defining the crystal structure without using a set of matched patterns.  Based on my work all ready, a backwards appreach  it should be possible to outputt a set of symmetries and the magnitude of their k-vectors. From there a backwards approach where you match the ratios and the symmetries the crystal structure and orientation should be (fairly) easy.

 **Regarding future implementations**
- In polar coordinate matching the sub pixel refinement for the zerobeam is already done in the polar reprojection.  Increasing the angular refinement would require more angular points. 
- If you could think of a way to do a modified 2-D correlation you might be able to have a very very small library.  
    - In one direction you are correlating along the azimuthal direction
    - In the other direction you are correlating along the FFT of the k axis (where the frequency corresponds to the interplanar spacing). (I'm not sure if this is completely possible but I think there might be a way to fit not only the rotation but also the spacing with a 2D FFT.

EirikOpheim commented:
@dnjohnstone and @CSSFrancis,

Thanks for your input! 

To follow up on your comments a bit:
Should I try and implement a _correlate2d function that can be called in the full frame correlation function? I suppose both the wrap and mask logic in the _correlate() function can be reused if it is slightly modified to fit the 2d case. 

> One thing that could be speed up is taking the fft of the image we are trying to index only once. This means that we don't have to calculate it over and over for each pattern.

Could the function head be something like:
_correlate2d( image_FT, template, axes = (0, 1), mask = None, wrap = True, Normalize = True),
where the image is already transformed as it is passed to the correlate2d function? I suppose this would be ideal for uses 1 and 2 mentioned by @dnjohnstone, but not so much for use 3. 

> For the sub pixel refinement, I think this basically the same as taking the image, subdividing the pixels and then running the correlation. If we subdivide correctly this might not be much slower if we are smart and get to a power of 2 so the FFT is fast.

If I understand you correctly, this would involve dividing every pixel into for instance 2x2 pixels, and then run the correlation again? I believe this would work, but then we could even skip the first correlation and go straight for the sub-pixel method with the desired pixel partition? 
In the article by A. Foden et al. I believe they have chosen a slightly different approach, where the initial correlation is used to locate the shift x_0, y_0 that gives the best correlation, and then an iterative method is used from there to find the local maximum for the correlation function using equations (5) and (6). 
I do, however, think that the method you mentioned of subdividing pixels would be both easy to implement and fast!

CSSFrancis commented:
@EirikOpheim sorry it's taken me a couple of days to respond,

> Should I try and implement a _correlate2d function that can be called in the full frame correlation function? I suppose both the wrap and mask logic in the _correlate() function can be reused if it is slightly modified to fit the 2d case.

That seems like a good place to start.  I'm  thinking that everything should be similar just expanded to 2 dimensions.  The masking stuff might get kind of complicated but I still think it is important for completeness sake. 

> > One thing that could be speed up is taking the fft of the image we are trying to index only once. This means that we don't have to calculate it over and over for each pattern.
> 
> Could the function head be something like:
> _correlate2d( image_FT, template, axes = (0, 1), mask = None, wrap = True, Normalize = True),
> where the image is already transformed as it is passed to the correlate2d function? I suppose this would be ideal for uses 1 and 2 mentioned by @dnjohnstone, but not so much for use 3.

I guess that would be an option.  It's going to end up looking kind of strange. Maybe we hold off on that and see if it needs to be faster.  Thinking about this more it makes things a little more complicated than necessary and might make things harder to maintain

> > For the sub pixel refinement, I think this basically the same as taking the image, subdividing the pixels and then running the correlation. If we subdivide correctly this might not be much slower if we are smart and get to a power of 2 so the FFT is fast.
> If I understand you correctly, this would involve dividing every pixel into for instance 2x2 pixels, and then run the correlation again? I believe this would work, but then we could even skip the first correlation and go straight for the sub-pixel method with the desired pixel partition?
> In the article by A. Foden et al. I believe they have chosen a slightly different approach, where the initial correlation is used to locate the shift x_0, y_0 that gives the best correlation, and then an iterative method is used from there to find the local maximum for the correlation function using equations (5) and (6).
> I do, however, think that the method you mentioned of subdividing pixels would be both easy to implement and fast!

I guess it would be something to try. [This]( https://dsp.stackexchange.com/questions/2321/is-up-sampling-prior-to-cross-correlation-useless) is a pretty good thread about upsampling before cross correlation that would be worth reading.  We are also helped because the simulated signals have infinite resolution. 

Hope that helps.

EirikOpheim commented:
I have now rewritten parts of the code using the next_fast_len() function. I also wrote a test for the new full_frame_correlation function. It seems like I also have to write a test for the correlate_library function for the case where method = "full_frame_correlation", but I cannot recall doing so for the zero_mean_normalized_correlation method. Also I cannot find any test for the correlate_library function. I would be very glad if someone has input on how to resolve this! 

Note: I decided against adding Fourier filter to the image within this function, as the exact same result can be achieved by applying a Fourier filter to the signal as a part of the preprocessing step (As pointed out by @hakonanes , I believe). And I believe such a filter function is better located in the diffraction2d class, along with other types of preprocessing. Do you guys agree with this?

dnjohnstone commented:
> I have now rewritten parts of the code using the next_fast_len() function. I also wrote a test for the new full_frame_correlation function. It seems like I also have to write a test for the correlate_library function for the case where method = "full_frame_correlation", but I cannot recall doing so for the zero_mean_normalized_correlation method. Also I cannot find any test for the correlate_library function. I would be very glad if someone has input on how to resolve this!

Thanks @EirikOpheim, sorry it took me a while to get to this, I think that @pc494 added the testing to your previous PR #545 for you? Anyway does this help: https://github.com/pyxem/pyxem/pull/545/commits/fa2f9eb5b0aab52bc2bd73a12726105339cf7765

> Note: I decided against adding Fourier filter to the image within this function, as the exact same result can be achieved by applying a Fourier filter to the signal as a part of the preprocessing step (As pointed out by @hakonanes , I believe). And I believe such a filter function is better located in the diffraction2d class, along with other types of preprocessing. Do you guys agree with this?

I think that Fourier filtering data is such standard thing that it should be done in workflow not with a build in method of the Diffraction2D class. I.e. our target user base should be able to write a bandpass filter and apply it using map in workflow so it shouldn't go in the code really. Do you agree @pc494 ?

Does that give you enough to finish this off @EirikOpheim ?

pc494 commented:
In general I agree with @dnjohnstone. 

@EirikOpheim if you could make a minimum working example that might help as I'm a little bit concerned that your modelling a diffraction pattern (rather than just template peak positions) under the hood here.

EirikOpheim commented:
> @EirikOpheim if you could make a minimum working example that might help as I'm a little bit concerned that your modelling a diffraction pattern (rather than just template peak positions) under the hood here.

I am modelling a diffraction pattern under the hood. I thought I would have to do so in order to do the Fourier transform. Is there an easy way to do so without defining a diffraction pattern, directly from the intensity list?

pc494 commented:
No, there probably isn't, but I would imagine you'll want to factorise the FFT out of the matching section. At present I think you'll end up performing M x N FFT's on templates (M experimental patterns and N templates) whereas you only need to do N of them.

EirikOpheim commented:
> No, there probably isn't, but I would imagine you'll want to factorise the FFT out of the matching section. At present I think you'll end up performing M x N FFT's on templates (M experimental patterns and N templates) whereas you only need to do N of them.

I see your point! I should definitely  factorise the FFT out. I am not quite sure how I should do this, though. Would it require the FFT to be performed on the template library in the correlate function, so that the fourier transformed templates are sent to the correlate library function? 
The code would be something like:

`if (method == 'Full_frame') :  \n
 FFT_Library = FFT(self.library) \n 
 matches = signal.map(  \n
 correlate_library, \n
 library=FFT_Library, ...)`

pc494 commented:
> > No, there probably isn't, but I would imagine you'll want to factorise the FFT out of the matching section. At present I think you'll end up performing M x N FFT's on templates (M experimental patterns and N templates) whereas you only need to do N of them.
> 
> I see your point! I should definitely factorise the FFT out. I am not quite sure how I should do this, though. Would it require the FFT to be performed on the template library in the correlate function, so that the fourier transformed templates are sent to the correlate library function?
> The code would be something like:
> 
> `if (method == 'Full_frame') : \n FFT_Library = FFT(self.library) \n matches = signal.map( \n correlate_library, \n library=FFT_Library, ...)`

Yep, I think you've got the same thing in mind as me.

EirikOpheim commented:
@pc494 I have made some changes to the structure now, following your input on doing the FFT for templates only once. I have had to restructure a bit, creating a new function correlate_library_from_dict, that takes in a dict structure in place of the template library for the correlate_library function. I have tried to make this function general enough to be applicable also to other methods that involve performing operations on the templates before comparison, such as the spherical decomposition method. Do you agree with the general structure I have chosen?

tjof2 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -21,6 +21,7 @@
 """
 
 import numpy as np
+from scipy.fftpack import next_fast_len
```
`scipy.fftpack` is the old module. I think something like this is better:
```python
def optimal_fft_size(target, real=False):
    """Wrapper around scipy function next_fast_len() for calculating optimal FFT padding.
    scipy.fft was only added in 1.4.0, so we fall back to scipy.fftpack
    if it is not available. The main difference is that next_fast_len()
    does not take a second argument in the older implementation.

    Parameters
    ----------
    target : int
        Length to start searching from. Must be a positive integer.
    real : bool, optional
        True if the FFT involves real input or output, only available
        for scipy > 1.4.0
    Returns
    -------
    int
        Optimal FFT size.
    """

    try:
        from scipy.fft import next_fast_len

        support_real = True

    except ImportError:
        from scipy.fftpack import next_fast_len

        support_real = False

    if support_real:
        return next_fast_len(target, real)
    else:
        return next_fast_len(target)
```
and then replace instances of `next_fast_len()` with `optimal_fft_size()`.

tjof2 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -113,28 +140,44 @@ def correlate(
+        elif method in ["full_frame_correlation"] :
+            shape = signal.data.shape[1:]
+            size = 2 * np.array(shape) - 1
+            fsize = [next_fast_len(a) for a in (size)]
```
If your data is real-valued/complex-valued, you can make use of the second argument.
```python
fsize = [optimal_fft_size(a, not signal.data.dtype.kind == 'c') for a in size]
```

tjof2 commented on pyxem/generators/indexation_generator.py:
```diff
@@ -45,6 +48,29 @@
+def get_fourier_transform(template_coordinates, template_intensities, shape, fsize):
+    template = np.zeros((shape))
+    template[template_coordinates[:, 1], template_coordinates[:, 0]] = template_intensities[:]
+    template_FT = np.fft.fftshift(np.fft.fftn(template,fsize))
```
Does it still work if you use `np.fft.rfftn()`? You get a 2x speed-up using that (assuming the data supports this assumption).

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -132,6 +133,144 @@ def zero_mean_normalized_correlation(
+
+    fprod = pattern_FT * image_FT
+
+    res_matrix = np.fft.ifftn(fprod)
```
Again, can you reliably use `np.fft.irfftn()` for the same result?

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -132,6 +133,144 @@ def zero_mean_normalized_correlation(
+    if method == "full_frame_correlation":
+        size = 2 * np.array(image.shape) - 1
+        fsize = [next_fast_len(a) for a in (size)]
+        image_FT = np.fft.fftshift(np.fft.fftn(image, fsize))
```
Same question with `np.fft.rfftn()`

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -205,6 +348,13 @@ def correlate_library(image, library, n_largest, method, mask):
+    if method == "full_frame_correlation":
+        size = 2 * np.array(image.shape) - 1
+        fsize = [next_fast_len(a) for a in (size)]
+        image_FT = np.fft.fftshift(np.fft.fftn(image, fsize))
```
And again `np.fft.rfftn()`

tjof2 replied:
Did you miss this one?

EirikOpheim replied:
Yes, fixed it now :)

tjof2 commented on pyxem/utils/indexation_utils.py:
```diff
@@ -205,6 +348,13 @@ def correlate_library(image, library, n_largest, method, mask):
+
+    if method == "full_frame_correlation":
+        size = 2 * np.array(image.shape) - 1
+        fsize = [next_fast_len(a) for a in (size)]
```
See above comment about `next_fast_len()` vs `optimal_fft_size()`

EirikOpheim commented:
@tjof2,

Thanks for the review! I have replaced np.fft.fftn() with np.fft.rfftn() everywhere in the code, as both the signal and the templates will have real values. I believe we have to use ifftn() and not irfftn() as the product of the Fourier transformed template and signal in general will be complex. 

I have added the wrapper function you wrote for next_fast_len(), and I am now using the wrapper function instead of using next_fast_len() directly.

pc494 commented:
Am I correct in thinking that a version of this is going to be PR'd to the 0.11.x branch at some point soon?

EirikOpheim commented:
> Am I correct in thinking that a version of this is going to be PR'd to the 0.11.x branch at some point soon?

Yes, I can make a new PR now :)
